### PR TITLE
initial attempt at alpine apk files

### PR DESCRIPTION
**Pull request by gdams**
fixes: https://github.com/adoptium/installer/issues/220

### CODE REVIEW

**Comment by lfvjimisola** [comment]
@gdams Thank you for your contribution. Many of us using alpine with be so grateful.

Out of curiosity, will the temurin jre/jdk alpine apk packages be hosted by Adoptium or by alpine?
Is the plan to be able to various temurin jre/jdk with different versions of alpine?

**Comment by gdams** [comment]
@lfvjimisola the apk files will be pushed to https://packages.adoptium.net (where we host our DEB/RPM files too)

I've been experimenting with pushing some apk files up if you want to have a go with them and provide some feedback?

add the public key:

```bash
wget -O /etc/apk/keys/adoptium.rsa.pub https://packages.adoptium.net/artifactory/api/security/keypair/public/repositories/apk
```

add the repo:

```bash
echo 'https://packages.adoptium.net/artifactory/apk/alpine/main' >> /etc/apk/repositories
```

install JDK (replace <version> with 17/11/8 etc)

```bash
apk update # if needed
apk add temurin-<version>-jdk
````

**Comment by gdams** [comment]
One problem I've observed is that we don't seem to be able to install multiple versions on the same machine:

```output
/ # apk add temurin-17-jdk
WARNING: This apk-tools is OLD! Some packages might not function properly.
ERROR: unable to select packages:
  temurin-17-jdk-17.0.4.1_p1-r0:
    conflicts: temurin-19-jdk-19.0.0_p36-r0[so:libattach.so=0] temurin-19-jdk-19.0.0_p36-r0[so:libawt.so=0]
               temurin-19-jdk-19.0.0_p36-r0[so:libawt_headless.so=0] temurin-19-jdk-19.0.0_p36-r0[so:libdt_socket.so=0]
               temurin-19-jdk-19.0.0_p36-r0[so:libextnet.so=0] temurin-19-jdk-19.0.0_p36-r0[so:libfontmanager.so=0]
               temurin-19-jdk-19.0.0_p36-r0[so:libinstrument.so=0] temurin-19-jdk-19.0.0_p36-r0[so:libj2gss.so=0]
               temurin-19-jdk-19.0.0_p36-r0[so:libj2pcsc.so=0] temurin-19-jdk-19.0.0_p36-r0[so:libj2pkcs11.so=0]
               temurin-19-jdk-19.0.0_p36-r0[so:libjaas.so=0] temurin-19-jdk-19.0.0_p36-r0[so:libjava.so=0]
               temurin-19-jdk-19.0.0_p36-r0[so:libjavajpeg.so=0] temurin-19-jdk-19.0.0_p36-r0[so:libjawt.so=0]
               temurin-19-jdk-19.0.0_p36-r0[so:libjdwp.so=0] temurin-19-jdk-19.0.0_p36-r0[so:libjimage.so=0]
               temurin-19-jdk-19.0.0_p36-r0[so:libjli.so=0] temurin-19-jdk-19.0.0_p36-r0[so:libjsig.so=0]
               temurin-19-jdk-19.0.0_p36-r0[so:libjsound.so=0] temurin-19-jdk-19.0.0_p36-r0[so:libjsvml.so=0]
               temurin-19-jdk-19.0.0_p36-r0[so:libjvm.so=0] temurin-19-jdk-19.0.0_p36-r0[so:liblcms.so=0]
               temurin-19-jdk-19.0.0_p36-r0[so:libmanagement.so=0] temurin-19-jdk-19.0.0_p36-r0[so:libmanagement_agent.so=0]
               temurin-19-jdk-19.0.0_p36-r0[so:libmanagement_ext.so=0] temurin-19-jdk-19.0.0_p36-r0[so:libmlib_image.so=0]
               temurin-19-jdk-19.0.0_p36-r0[so:libnet.so=0] temurin-19-jdk-19.0.0_p36-r0[so:libnio.so=0]
               temurin-19-jdk-19.0.0_p36-r0[so:libprefs.so=0] temurin-19-jdk-19.0.0_p36-r0[so:librmi.so=0]
               temurin-19-jdk-19.0.0_p36-r0[so:libsaproc.so=0] temurin-19-jdk-19.0.0_p36-r0[so:libsctp.so=0]
               temurin-19-jdk-19.0.0_p36-r0[so:libsyslookup.so=0] temurin-19-jdk-19.0.0_p36-r0[so:libverify.so=0]
               temurin-19-jdk-19.0.0_p36-r0[so:libzip.so=0]
    satisfies: world[temurin-17-jdk]
  temurin-19-jdk-19.0.0_p36-r0:
    conflicts: temurin-17-jdk-17.0.4.1_p1-r0[so:libattach.so=0] temurin-17-jdk-17.0.4.1_p1-r0[so:libawt.so=0]
               temurin-17-jdk-17.0.4.1_p1-r0[so:libawt_headless.so=0] temurin-17-jdk-17.0.4.1_p1-r0[so:libdt_socket.so=0]
               temurin-17-jdk-17.0.4.1_p1-r0[so:libextnet.so=0] temurin-17-jdk-17.0.4.1_p1-r0[so:libfontmanager.so=0]
               temurin-17-jdk-17.0.4.1_p1-r0[so:libinstrument.so=0] temurin-17-jdk-17.0.4.1_p1-r0[so:libj2gss.so=0]
               temurin-17-jdk-17.0.4.1_p1-r0[so:libj2pcsc.so=0] temurin-17-jdk-17.0.4.1_p1-r0[so:libj2pkcs11.so=0]
               temurin-17-jdk-17.0.4.1_p1-r0[so:libjaas.so=0] temurin-17-jdk-17.0.4.1_p1-r0[so:libjava.so=0]
               temurin-17-jdk-17.0.4.1_p1-r0[so:libjavajpeg.so=0] temurin-17-jdk-17.0.4.1_p1-r0[so:libjawt.so=0]
               temurin-17-jdk-17.0.4.1_p1-r0[so:libjdwp.so=0] temurin-17-jdk-17.0.4.1_p1-r0[so:libjimage.so=0]
               temurin-17-jdk-17.0.4.1_p1-r0[so:libjli.so=0] temurin-17-jdk-17.0.4.1_p1-r0[so:libjsig.so=0]
               temurin-17-jdk-17.0.4.1_p1-r0[so:libjsound.so=0] temurin-17-jdk-17.0.4.1_p1-r0[so:libjsvml.so=0]
               temurin-17-jdk-17.0.4.1_p1-r0[so:libjvm.so=0] temurin-17-jdk-17.0.4.1_p1-r0[so:liblcms.so=0]
               temurin-17-jdk-17.0.4.1_p1-r0[so:libmanagement.so=0] temurin-17-jdk-17.0.4.1_p1-r0[so:libmanagement_agent.so=0]
               temurin-17-jdk-17.0.4.1_p1-r0[so:libmanagement_ext.so=0] temurin-17-jdk-17.0.4.1_p1-r0[so:libmlib_image.so=0]
               temurin-17-jdk-17.0.4.1_p1-r0[so:libnet.so=0] temurin-17-jdk-17.0.4.1_p1-r0[so:libnio.so=0]
               temurin-17-jdk-17.0.4.1_p1-r0[so:libprefs.so=0] temurin-17-jdk-17.0.4.1_p1-r0[so:librmi.so=0]
               temurin-17-jdk-17.0.4.1_p1-r0[so:libsaproc.so=0] temurin-17-jdk-17.0.4.1_p1-r0[so:libsctp.so=0]
               temurin-17-jdk-17.0.4.1_p1-r0[so:libsyslookup.so=0] temurin-17-jdk-17.0.4.1_p1-r0[so:libverify.so=0]
               temurin-17-jdk-17.0.4.1_p1-r0[so:libzip.so=0]
    satisfies: world[temurin-19-jdk]
```

I'm not massively familiar with the APKBUILD syntax so I'll have to do some reading up to understand what's going wrong here.

**Comment by gdams** [comment]
asked the question at https://gitlab.alpinelinux.org/alpine/abuild/-/issues/10082

**Comment by gdams** [comment]
> One problem I've observed is that we don't seem to be able to install multiple versions on the same machine:

This is now fixed

**Comment by zdtsw** [comment]
in all alpine's logs: `ERROR: temurin-X-jdk*: libasound.so.2: path not found` is the depedency track failing?

**Comment by gdams** [comment]
> in all alpine's logs: `ERROR: temurin-X-jdk*: libasound.so.2: path not found` is the depedency track failing? 

Hmm yea looks like I need to add alsa-lib as a dependency.

**Comment by gdams** [comment]
@zdtsw I've fixed that error, there is still a warning about `libawt.so` for JDK8u but that's bundled as part of the JDK so likely not a concern

**Review by zdtsw — APPROVED** [review]

**Review by karianna — APPROVED** [review]
LGTM

File: linux/jdk/alpine/build.gradle [hunk 1]
```diff
@@ -0,0 +1,143 @@
+plugins {
+	id "java"
+}
+
+ext {
+	junitVersion = "5.9.0"
+	testcontainersVersion = "1.17.3"
```
Comment (karianna) [inline linux/jdk/alpine/build.gradle:7]: 1.17.5 now

File: linux/jdk/alpine/build.gradle [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+	assertjCoreVersion = "3.23.1"
+}
+
+repositories {
+	mavenCentral()
+}
+
+group "org.adoptium"
+version "1.0.0-SNAPSHOT"
```
Comment (karianna) [inline linux/jdk/alpine/build.gradle:16]: should that go to 1.0.0?

**Review by karianna — APPROVED** [review]

File: linux/jdk/alpine/src/main/packaging/temurin/11/APKBUILD [hunk 3]
```diff
@@ -0,0 +1,96 @@
+# Maintainer: Eclipse Adoptium Package Maintainers <temurin-dev@eclipse.org>
+pkgname=temurin-11
+pkgver=11.0.16.1_p1
```
[the remark this answers is not in the document]
Comment (gdams) [reply linux/jdk/alpine/src/main/packaging/temurin/11/APKBUILD:3]: yes - same for the other packages

File: linux/jdk/alpine/build.gradle [hunk 4]
```diff
@@ -0,0 +1,143 @@
+plugins {
+	id "java"
+}
+
+ext {
+	junitVersion = "5.9.0"
```
[the remark this answers is not in the document]
Comment (gdams) [reply linux/jdk/alpine/build.gradle:6]: I'd prefer to not add dependency changes in this PR. We should add dependabot to handle the gradle updates

File: linux/jdk/alpine/build.gradle (hunk 2 above)
Comment (gdams) [reply linux/jdk/alpine/build.gradle:16]: I'd prefer to not add dependency changes in this PR. We should add dependabot to handle the gradle updates

File: linux/jdk/alpine/build.gradle (hunk 1 above)
Comment (gdams) [reply linux/jdk/alpine/build.gradle:7]: I'd prefer to not add dependency changes in this PR. We should add dependabot to handle the gradle updates

**Comment by sophia-guo** [comment]
Would it be better to move HelloWorld.java, TestCryptoLevel.java and TestECDSA.java to PackageTest 
 folder instead of packaging folder as those are tests? And feels like there is no need to keep one copy for each version?

**Comment by zdtsw** [comment]
> Would it be better to move HelloWorld.java, TestCryptoLevel.java and TestECDSA.java to PackageTest folder instead of packaging folder as those are tests? And feels like there is no need to keep one copy for each version?

i think the difference for these are they are testing for different purposes.
the HelloWolrd and so, are part of the apkbuild check() to use the just-built-jdk to compile java code and run it as some basic sanity checks.
the others we are running like rpm test are using testcontainer which is running a bunch of java testcases  to test if the final package. so the jdk for those tests are not the ones we intended to release.

**Comment by gdams** [comment]
test code freeze

**Comment by lfvjimisola** [comment]
> gdams

We are using it for one of our containers that is not used in production yet. It works well from what we have seen so far.
Any idea, on when these apk will be added to the official repo.

Thanks a lot for your work!

**Comment by lfvjimisola** [comment]
@gdams It no longer works for us with the steps below. Has there been any chances now that your contribution has closed the issue?

```
RUN curl https://packages.adoptium.net/artifactory/api/security/keypair/public/repositories/apk -o /etc/apk/keys/adoptium.rsa.pub \ 
    echo 'https://packages.adoptium.net/artifactory/apk/alpine/main' >> /etc/apk/repositories \
    apk update && apk add temurin-17-jdk
```

**Pull request merged**